modules/multisignatures.js

```diff
@@ -80,158 +80,221 @@ function Multisignatures(cb, scope) {
 
-			if (multisignatureTransaction.requesterPublicKey) {
-				multisignatures.push(multisignatureTransaction.senderPublicKey);
-			}
+			// Emit events
+			library.network.io.sockets.emit(
+				'multisignatures/signature/change',
+				transaction
+			);
+			library.bus.message('signature', signature, true);
+			return setImmediate(cb);
+		}
+	);
+};
 
-			if (!account) {
-				return setImmediate(cb, 'Account not found');
-			}
+/**
+ * Main function for processing received signature, includes:
+ * - multisignature account creation
+ * - send from multisignature account
+ *
+ * @public
+ * @param {Object} signature - Signature data
+ * @param {string} [signature.publicKey] - Public key of account that created the signature (optional)
+ * @param {string} signature.transactionId - Id of transaction that signature was created for
+ * @param {string} signature.signature - Actual signature
+ * @param {function} cb - Callback function
+ * @implements {library.balancesSequence.add} - All processing here is done through balancesSequence
+ * @returns {setImmediateCallback} cb, err
+ */
+Multisignatures.prototype.processSignature = function(signature, cb) {
+	if (!signature) {
+		const message = 'Unable to process signature, signature not provided';
+		library.logger.error(message);
+		return setImmediate(cb, message);
+	}
 
-			multisignatureTransaction.signatures =
-				multisignatureTransaction.signatures || [];
+	// From now perform all the operations via balanceSequence
+	library.balancesSequence.add(balanceSequenceCb => {
+		// Grab transaction with corresponding ID from transaction pool
+		const transaction = modules.transactions.getMultisignatureTransaction(
+			signature.transactionId
+		);
 
-			if (
-				multisignatureTransaction.signatures.indexOf(transaction.signature) >= 0
-			) {
-				return setImmediate(cb, 'Signature already exists');
-			}
+		if (!transaction) {
+			const message =
+				'Unable to process signature, corresponding transaction not found';
+			library.logger.error(message, { signature });
+			return setImmediate(balanceSequenceCb, message);
+		}
 
-			try {
-				for (let i = 0; i < multisignatures.length && !verify; i++) {
-					verify = library.logic.transaction.verifySignature(
-						multisignatureTransaction,
-						multisignatures[i],
-						transaction.signature
-					);
-				}
-			} catch (e) {
-				library.logger.error(e.stack);
-				return setImmediate(cb, 'Failed to verify signature');
-			}
+		// If there are no signatures yet - initialise with empty array
+		transaction.signatures = transaction.signatures || [];
 
-			if (!verify) {
-				return setImmediate(cb, 'Failed to verify signature');
-			}
+		// Check if received signature already exists in transaction
+		if (transaction.signatures.includes(signature.signature)) {
+			const message = 'Unable to process signature, signature already exists';
+			library.logger.error(message, { signature, transaction });
```

Comment:
Is signature and transactions logged here as an actual object or 'Object [Object]'

Reply:
Not sure, will check it, thanks.

Reply:
Did you check?

Reply:
Yes, it's fine:
```
[ERR] 2018-08-02 23:28:25 | Unable to process signature, signature already exists - {"signature":{"transactionId":"6242316680238493239","publicKey":"7782833dde90339776320cd765d1acbacb46e2cda060bc8b0b065a8f0579abe1","signature":"8ceae235015aca221ce888395b6770972f7eca39485f75674fa72682b9b751604c395ffd0afb9e27862fb85e2fdab1a179d8264769e8d6e77f55d391dd613c0f"},"transaction":{"amount":"0","senderPublicKey":"7ddf8cdcf374caa99c7a12051458534c0753916762bc0c47c75019e4997385e4","timestamp":69143266,"type":4,"fee":"1500000000","asset":{"multisignature":{"min":2,"lifetime":1,"keysgroup":["+7782833dde90339776320cd765d1acbacb46e2cda060bc8b0b065a8f0579abe1","+d88800d9290b595a1b692b29c91759e49c64d41b2fbae58b745f84297a92834c"]}},"signature":"e22ff1f54f7f89b6276aace8cee8c4098833f88a5c3e4610cc5051b124e0564e9fc3a4a5e8b5521cdd73cd320cb866a304dd7a54950d21bd727188df06e43202","id":"6242316680238493239","senderId":"16150081938385139566L","relays":1,"receivedAt":"2018-08-02T23:28:24.853Z","signatures":["8ceae235015aca221ce888395b6770972f7eca39485f75674fa72682b9b751604c395ffd0afb9e27862fb85e2fdab1a179d8264769e8d6e77f55d391dd613c0f"],"ready":false}}
[ERR] 2018-08-02 23:28:25 | Unable to process signature, signer not in keysgroup. - {"signature":{"transactionId":"6242316680238493239","publicKey":"3f0f1aab87667899844ae92c4a80b2a795396e3cef9d49f766bef1222946ea7d","signature":"7e84bb8dd2fee2ae9cf254ae679e136839a9f47e40ca760c6fcc769b57c7c4844973bb70a85a987bf8c2b990f43268e2735e28bfe5a156c055afd475b85a340a"},"membersPublicKeys":["7782833dde90339776320cd765d1acbacb46e2cda060bc8b0b065a8f0579abe1","d88800d9290b595a1b692b29c91759e49c64d41b2fbae58b745f84297a92834c"],"transaction":{"amount":"0","senderPublicKey":"7ddf8cdcf374caa99c7a12051458534c0753916762bc0c47c75019e4997385e4","timestamp":69143266,"type":4,"fee":"1500000000","asset":{"multisignature":{"min":2,"lifetime":1,"keysgroup":["+7782833dde90339776320cd765d1acbacb46e2cda060bc8b0b065a8f0579abe1","+d88800d9290b595a1b692b29c91759e49c64d41b2fbae58b745f84297a92834c"]}},"signature":"e22ff1f54f7f89b6276aace8cee8c4098833f88a5c3e4610cc5051b124e0564e9fc3a4a5e8b5521cdd73cd320cb866a304dd7a54950d21bd727188df06e43202","id":"6242316680238493239","senderId":"16150081938385139566L","relays":1,"receivedAt":"2018-08-02T23:28:24.853Z","signatures":["8ceae235015aca221ce888395b6770972f7eca39485f75674fa72682b9b751604c395ffd0afb9e27862fb85e2fdab1a179d8264769e8d6e77f55d391dd613c0f"],"ready":false}}
```